pkg/querysharding/analyzer.go

```diff
@@ -6,21 +6,65 @@ package querysharding
 import (
 	"fmt"
 
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/prometheus/prometheus/promql/parser"
 )
 
 // QueryAnalyzer is an analyzer which determines
 // whether a PromQL Query is shardable and using which labels.
+
+type Analyze interface {
+	Analyze(string) (QueryAnalysis, error)
+}
+
 type QueryAnalyzer struct{}
 
+type CachedQueryAnalyzer struct {
+	analyzer *QueryAnalyzer
+	cache    *lru.Cache
+}
+
 var nonShardableFuncs = []string{
 	"label_join",
 	"label_replace",
 }
 
 // NewQueryAnalyzer creates a new QueryAnalyzer.
-func NewQueryAnalyzer() *QueryAnalyzer {
-	return &QueryAnalyzer{}
+func NewQueryAnalyzer() (Analyze, error) {
```

Comment:
I prefer to return a concrete type like `*CachedQueryAnalyzer` rather than the interface